spec.md

```diff
@@ -0,0 +1,81 @@
+
+#### Tracer
+
+The `Tracer` interface must have the following capabilities:
+
+- Start a `Span` that has no parent **(py: `start_trace`, go: `StartTrace`)**
+- Start a `Span` given a parent `TraceContext` **(py: `join_trace`, go: `JoinTrace`)**
+- Start a `Span` associated with a specific `TraceContext` **(py: `start_span_with_context`, go: `StartSpanWithContext`)**
+
+
+#### Span
+
+The `Span` interface must have the following capabilities:
+
+- Create and start a new child `Span` with a given operation name **(py: `start_child`, go: `StartChild`)**
+- Finish the (already-started) `Span`.  Finish should be the last call made to any span instance, and to do otherwise leads to undefined behavior. **(py: `finish`, go: `Finish`)**
+- Set a key:value tag on the `Span`. The key must be a `string`, and the value must be either a `string`, a `boolean`, or a numeric type. Behavior for other value types is undefined. If multiple values are set to the same key (i.e., in multiple calls), implementation behavior is also undefined. **(py: `set_tag`, go: `SetTag`)**
+- Add a new info `Log` to the `Span`, accepting a message `string` and 0 or more payload arguments. The payload arguments may be of any type and arbitrary size, though implementations are not required to retain all payload arguments (or even all parts of all payload arguments). **(py: `info`, go: `Info`)**
+- Add a new error `Log` to the `Span`; analagous to adding an info log aside from the severity. **(py: `error`, go: `Error`)**
+- Access the `TraceContext` associated with the `Span`. **(py: `trace_context`, go: `TraceContext()`)**
+
+
+#### TraceContext
+
+Every `TraceContext` must provide access to "trace attributes". A **trace attribute** is a key:value pair associated with a TraceContext **that also propagates to future TraceContext children** per TraceContext.NewChild, and that propagates across process boundaries in-band with application data.
+
+Trace attributes are powerful, especially given an OpenTracing integration that extends across many layers in a distributed stack. For example, arbitrary application data from a mobile app can travel, transparently, all the way into the depths of a storage system). These powerful capabilities come with some powerful costs: use this feature with care.
+
+Also, trace attribute keys have a restricted format: implementations may wish to use them as HTTP header keys (or key suffixes), and of course HTTP headers are case insensitive.
+
+As such, trace attribute keys MUST match the regular expression `(?i:[a-z0-9][-a-z0-9]*)`, and – per the `?i:` – they are case-insensitive.  That is, the trace attribute key must start with a letter or number, and the remaining characters must be letters, numbers, or hyphens. See CanonicalizeTraceAttributeKey(). If a trace attribute key does not meet these criteria, SetTraceAttribute() results in undefined behavior.
```

Comment:
Since it's a spec / language independent, we shouldn't reference an implementation `CanonicalizeTraceAttributeKey` here. I think it'd be fine to just remove that part.